File: cpp/032最长有效括号/032最长有效括号.cpp
/*
给定一个只包含 '(' 和 ')' 的字符串，找出最长的包含有效括号的子串的长度。

示例 1:

输入: "(()"
输出: 2
解释: 最长有效括号子串为 "()"
示例 2:

输入: ")()())"
输出: 4
解释: 最长有效括号子串为 "()()"
*/

//解答： 二重循环，每次选择一个字符作为起始点，如果右括号增加后使得右括号左括号数目达到一致，则为有效括号串，更新当前子串长度，如果右括号增加后数量大于左括号数，则不合法跳出循环
//比较每个子循环记录的长度和最大长度，如果大于最大长度则更新
#include <iostream>
#include <string>
using namespace std;
class Solution {
public:
    int longestValidParentheses(string s) {
      if(s.length() == 0)
          return 0;
      int maxLength = 0;
      int currentLength;
      int j;
      int left;
      int right;
      for(int i = 0; i < s.length(); i++)
      {
          left = 0;
          right = 0;
          currentLength = 0;
          for(j = i; j < s.length(); j++)
          {
               if(s[j] == '(')
               {
                    left++;
               }
              else
              {
                   right++;
                  if(right > left)
                  {
                      break;
                  }
                  else if(right == left)
                  {
                      currentLength = j - i + 1;
                  }
              }
              
          }
          
          if(currentLength > maxLength)
        {
            maxLength = currentLength;          
        }
      }
        
        return maxLength;
    }
};

File: python/032最长有效括号/032最长有效括号.py
"""
给定一个只包含 '(' 和 ')' 的字符串，找出最长的包含有效括号的子串的长度。

示例 1:

输入: "(()"
输出: 2
解释: 最长有效括号子串为 "()"
示例 2:

输入: ")()())"
输出: 4
解释: 最长有效括号子串为 "()()"
"""
# 解答：用动态规划做，详见下面注释

class Solution:
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        f = [0] * n # 动态规划辅助函数，每个元素表示以当前下标开始往后的有效括号子串长度
        maxLength = 0 # 记录最长长度
        for i in range(n - 2, -1, -1):  # 逆序遍历
            if s[i] == '(': # 当前为左括号
                # i + 1为下一个元素，f[i+1]为下一个元素的有效子串长度，所以i + 1 + f[i + 1] 为跳过已经有效的子串，看后面一个是否为右括号
                if i + 1  + f[i + 1] < n and s[i + 1 + f[i + 1]] == ')' :
                    f[i] = 2 + f[i + 1]
                # 处理连续相邻有效子串的情况,即如果相邻的后面还是一个有效括号串，则也需要加入
                    if i + 1 + f[i + 1] + 1 < n:
                        f[i] = f[i] + f[i + 1 + f[i + 1] + 1]
                    if f[i] > maxLength:
                        maxLength = f[i]
                        
        return maxLength
            